src/main/scala/org/broadinstitute/gpp/poolq3/reports/UnexpectedSequenceWriter.scala

```diff
@@ -52,96 +50,123 @@ object UnexpectedSequenceWriter {
     tryDelete(unexpectedSequenceCacheDir)
   }
 
-  // builds the set of unexpected sequences we will actually count data for
-  private[reports] def sampleCache(
-    cacheDir: Path,
-    samplePct: Double,
-    colReference: Reference,
-    unexpectedCountsByBarcode: Map[String, Int]
-  ): Set[String] = {
-    // this set will contain the barcodes to track for all shards
-    val ret = mutable.HashSet[String]()
-
-    // read the shard for each column barcode in turn
-    colReference.allBarcodes.foreach { dnaBarcode =>
-      // compute the number of lines to read for this shard by the samplePct and the number of reads we got for it
-      val linesToRead: Int = math.ceil(unexpectedCountsByBarcode.getOrElse(dnaBarcode, 0) * samplePct).toInt
-      val file = cacheDir.resolve(nameFor(dnaBarcode))
-      if (Files.exists(file)) {
-        // add all the row barcodes found in the first `linesToRead` lines to `ret`
-        Using.resource(Source.fromFile(file.toFile)) { src =>
-          src.getLines().take(linesToRead).foreach(line => ret += line)
+  // defining this as a trait is not useful in the main codebase but it makes testing easier
+  private[reports] trait CachedBarcodes extends Iterator[String] with Closeable {
+    def colBc: String
+  }
+
+  final private[reports] class SourceCachedBarcodes(val colBc: String, source: Source) extends CachedBarcodes {
+    private val iter = source.getLines()
+    def close(): Unit = source.close()
+    def hasNext: Boolean = iter.hasNext
+    def next(): String = iter.next()
+  }
+
+  final private[reports] class BreadthFirstIterator(readers: mutable.ArrayBuffer[CachedBarcodes])
+      extends Iterator[(String, String)] {
+    private var i = 0
+
+    // this only works if all the readers were non-empty at the start
+    def hasNext: Boolean = readers.nonEmpty && readers(i).hasNext
+
+    def next(): (String, String) = {
```

Comment:
Tricky stuff but it looks like it works!   Did you entertain the idea of a dequeue where you pop a reader from the front and then push it back on the back only if it's nonempty?